### PR TITLE
Feature - Search - Implement Search

### PR DESCRIPTION
**Pull request by at-tamnguyen7**
#### Descriptions (Required)
- [ ] Describe about this pull request.

#### References (Optional)
- [ ] Link to the feature/bug. (if has)

#### Screenshot (Optional)
- [ ] UI pull request must contain this section.

### CODE REVIEW

File: FinalProject/Model/API/API.Covid.swift [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+//
+//  Created by PCI0001 on 9/21/20.
+//  Copyright © 2020 Thinh Nguyen X. All rights reserved.
+//
+
+import Foundation
+import Alamofire
+import  ObjectMapper
+
+
```
Comment (at-thinhnguyen-2) [inline FinalProject/Model/API/API.Covid.swift:13]: WARNING - Limit vertical whitespace to a single empty line. Currently 2.

Comment (at-thinhnguyen-2) [inline FinalProject/Model/API/API.Covid.swift:13]: WARNING - Files should have a single trailing newline.

File: FinalProject/Model/API/Core/Api.swift [hunk 2]
```diff
@@ -20,6 +20,27 @@ final class Api {
     }
 }
 
+final class Path {
+
```
Comment (at-thinhnguyen-2) [inline FinalProject/Model/API/Core/Api.swift:24]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/API/Core/Api.swift [hunk 3]
```diff
@@ -20,6 +20,27 @@ final class Api {
...
+final class Path {
+    
+    struct CovidPath {
+        static let baseURL = "https://api.covid19api.com"
+        static let summaryURL = "/summary"
+        static let liveURL = "/live/country"
+        static let worldURL = "/world/total"
+        static let countriesURL = "/countries"
+    }
+
```
Comment (at-thinhnguyen-2) [inline FinalProject/Model/API/Core/Api.swift:32]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/API/Core/ApiManager.swift [hunk 4]
```diff
@@ -13,7 +13,25 @@ typealias JSObject = [String: Any]
...
+typealias APICompleetion = (APIResult) -> Void
 
+enum APIResult {
+    case success
+    case failure(Error)
+}
+
+// MARK: - Get error for api result
+extension APIResult {
+
```
Comment (at-thinhnguyen-2) [inline FinalProject/Model/API/Core/ApiManager.swift:25]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/Schema/Country.swift [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+//
+//  Created by PCI0001 on 9/17/20.
+//  Copyright © 2020 Thinh Nguyen X. All rights reserved.
+//
+
+import Foundation
+import ObjectMapper
+
+final class Country: Mappable {
+
```
Comment (at-thinhnguyen-2) [inline FinalProject/Model/Schema/Country.swift:13]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/Schema/Country.swift [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    // MARK: - Properties
+    var countryName: String?
+    var countryCode: String?
+    var newConfirmed: Int?
+    var totalconfirmed: Int?
+    var newDeaths: Int?
+    var totalDeaths: Int?
+    var newRecovered: Int?
+    var totalRecovered: Int?
+
```
Comment (at-thinhnguyen-2) [inline FinalProject/Model/Schema/Country.swift:23]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/Schema/Country.swift [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    var totalconfirmed: Int?
+    var newDeaths: Int?
+    var totalDeaths: Int?
+    var newRecovered: Int?
+    var totalRecovered: Int?
+    
+    // MARK: - Initial
+    init?(map: Map) {
+    }
+
```
Comment (at-thinhnguyen-2) [inline FinalProject/Model/Schema/Country.swift:27]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/Schema/Covid.swift [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+//
+//  Created by PCI0001 on 9/17/20.
+//  Copyright © 2020 Thinh Nguyen X. All rights reserved.
+//
+
+import Foundation
+import ObjectMapper
+
+final class Covid: Mappable {
+
```
Comment (at-thinhnguyen-2) [inline FinalProject/Model/Schema/Covid.swift:13]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/Schema/Covid.swift [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+//
+
+import Foundation
+import ObjectMapper
+
+final class Covid: Mappable {
+    
+    // MARK: - Properties
+    var countries: [Country]?
+
```
Comment (at-thinhnguyen-2) [inline FinalProject/Model/Schema/Covid.swift:16]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/Schema/Covid.swift [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+final class Covid: Mappable {
+    
+    // MARK: - Properties
+    var countries: [Country]?
+    
+    // MARK: - Initial
+    init?(map: Map) {
+    }
+
```
Comment (at-thinhnguyen-2) [inline FinalProject/Model/Schema/Covid.swift:20]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/Schema/Global.swift [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+//
+//  Created by PCI0001 on 9/17/20.
+//  Copyright © 2020 Thinh Nguyen X. All rights reserved.
+//
+
+import Foundation
+import ObjectMapper
+
+final class Global: Mappable {
+
```
Comment (at-thinhnguyen-2) [inline FinalProject/Model/Schema/Global.swift:13]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/Schema/Global.swift [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+final class Global: Mappable {
+    
+    // MARK: - Properties
+    var newConfirmed: Int?
+    var totalconfirmed: Int?
+    var newDeaths: Int?
+    var totalDeaths: Int?
+    var newRecovered: Int?
+    var totalRecovered: Int?
+
```
Comment (at-thinhnguyen-2) [inline FinalProject/Model/Schema/Global.swift:21]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/Model/Schema/Global.swift [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    var totalconfirmed: Int?
+    var newDeaths: Int?
+    var totalDeaths: Int?
+    var newRecovered: Int?
+    var totalRecovered: Int?
+    
+    // MARK: - Initial
+    init?(map: Map) {
+    }
+
```
Comment (at-thinhnguyen-2) [inline FinalProject/Model/Schema/Global.swift:25]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/View/Search/Cell/SearchCellModel.swift [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+    var totalDeadth: Int? {
+        return country.totalDeaths
+    }
+    var totalRecovered: Int? {
+        return country.totalRecovered
+    }
+    var countryCode: String? {
+        return country.countryCode
+    }
+
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/Cell/SearchCellModel.swift:30]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/View/Search/Cell/SearchTableViewCell.swift [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+
+final class SearchTableViewCell: UITableViewCell {
+
+    // MARK: - IBOulets
+    @IBOutlet private weak var countryNameLabel: UILabel!
+    @IBOutlet private weak var countryCodeLabel: UILabel!
+    @IBOutlet private weak var numbercaseLabel: UILabel!
+    @IBOutlet private weak var numberDeadthLabel: UILabel!
+    @IBOutlet private weak var numberRecoveredLabel: UILabel!
+
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/Cell/SearchTableViewCell.swift:19]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/View/Search/Cell/SearchTableViewCell.swift [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    @IBOutlet private weak var numberDeadthLabel: UILabel!
+    @IBOutlet private weak var numberRecoveredLabel: UILabel!
+    
+    // MARK: - Properties
+    var viewModel: SearchCellModel? {
+        didSet {
+            updateView()
+        }
+    }
+
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/Cell/SearchTableViewCell.swift:26]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/View/Search/SearchViewController.swift [hunk 17]
```diff
@@ -8,23 +8,42 @@
 
 import UIKit
 
-class SearchViewController: UIViewController {
-
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/SearchViewController.swift:12]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/View/Search/SearchViewController.swift [hunk 18]
```diff
@@ -8,23 +8,42 @@
...
+final class SearchViewController: ViewController {
+    
+    // MARK: - IBOulets
+    @IBOutlet private weak var tableView: UITableView!
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/SearchViewController.swift:14]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/View/Search/SearchViewController.swift [hunk 19]
```diff
@@ -8,23 +8,42 @@
...
+    @IBOutlet private weak var searchBar: UISearchBar!
+    
+    // MARK: - Properties
+    private var cellIdentifier: String = "SearchCell"
+    // MARK: - Life cycle
     override func viewDidLoad() {
         super.viewDidLoad()
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/SearchViewController.swift:21]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/View/Search/SearchViewController.swift [hunk 20]
```diff
@@ -8,23 +8,42 @@
...
     }
-
-
-    /*
-    // MARK: - Navigation
-
-    // In a storyboard-based application, you will often want to do a little preparation before navigation
-    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
-        // Get the new view controller using segue.destination.
-        // Pass the selected object to the new view controller.
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/SearchViewController.swift:26]: WARNING - Lines should not have trailing whitespace.

File: FinalProject/View/Search/SearchViewController.swift [hunk 21]
```diff
@@ -8,23 +8,42 @@
...
+    
+    override func setUpUI() {
+        let nib = UINib(nibName: "SearchTableViewCell", bundle: .main)
+        tableView.register(nib, forCellReuseIdentifier: cellIdentifier)
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/SearchViewController.swift:26]: WARNING - Lines should not have trailing whitespace.

**Comment by at-thinhnguyen-2** [comment]
Total linter violations in pull request: 22
Serious: 0
Warnings: 22

File: FinalProject/View/Search/SearchViewModel.swift [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+
+final class SearchViewModel {
+    
+    // MARK: - Properties
+    var countries: [Country] = []
+    var filter: [Country] = []
+    
+    // MARK: - Function
+    func getData(completion: @escaping APICompletion) {
+        Api.Path.Search.getAllDataInSearch() { (result) in
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/SearchViewModel.swift:19]: WARNING - When using trailing closures, empty parentheses should be avoided after the method call.

File: FinalProject/View/Search/SearchViewController.swift [hunk 23]
```diff
@@ -8,23 +8,101 @@
...
+        tableView.delegate = self
+        tableView.dataSource = self
+    }
+    // MARK: - Private methods
+    private func configSearchController() {
+        title = "Search"
+        navigationItem.searchController = searchController
+        searchController.obscuresBackgroundDuringPresentation = false
+        searchController.searchResultsUpdater = self
+        searchController.searchBar.placeholder = "Type country here"
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/SearchViewController.swift:40]: WARNING - When using trailing closures, empty parentheses should be avoided after the method call.

File: FinalProject/View/Search/SearchViewController.swift [hunk 24]
```diff
@@ -8,23 +8,101 @@
...
+                DispatchQueue.main.async {
+                    self.alert(error: error)
+                }
+            }
+        }
+    }
+    
+   private func filterCountry(for searchText: String) {
+            if searchText.isEmpty {
+                viewModel.filter = viewModel.countries
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/SearchViewController.swift:60]: ERROR - Force unwrapping should be avoided.

File: FinalProject/View/Search/SearchViewController.swift [hunk 25]
```diff
@@ -8,23 +8,42 @@
...
-        // Pass the selected object to the new view controller.
+    
+    override func setUpUI() {
+        let nib = UINib(nibName: "SearchTableViewCell", bundle: .main)
+        tableView.register(nib, forCellReuseIdentifier: cellIdentifier)
+        tableView.delegate = self
+        tableView.dataSource = self
     }
-    */
+    // MARK: - Private methods
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/SearchViewController.swift:34]: WARNING - When using trailing closures, empty parentheses should be avoided after the method call.

**Comment by at-thinhnguyen-2** [comment]
Total linter violations in pull request: 44
Serious: 0
Warnings: 43

**Comment by at-thinhnguyen-2** [comment]
Total linter violations in pull request: 2
Serious: 0
Warnings: 2

File: FinalProject/Model/API/API.Covid.swift [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+            }
+        }
+    }
+}
+
+extension Api.Path.Detail {
+
+        static func getDataCellOne(completion: @escaping CompletionResult<Country>) {
+            let urlString = path
+            api.request(method: .get, urlString: urlString) { (result) in
```
Comment (at-thinhnguyen-2) [inline FinalProject/Model/API/API.Covid.swift:41]: WARNING - Unused parameter "result" in a closure should be replaced with _.

Comment (at-thinhnguyen-2) [inline FinalProject/Model/API/API.Covid.swift:41]: WARNING - Unused parameter "result" in a closure should be replaced with _.

File: FinalProject/Model/API/API.Covid.swift [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+                    }
+                } else {
+                    completion( .failure(Api.Error.emptyData))
+                }
+            case .failure(let error):
+                completion( .failure(error))
+            }
+        }
+    }
+}
```
Comment (at-thinhnguyen-2) [inline FinalProject/Model/API/API.Covid.swift:35]: WARNING - Unused parameter "result" in a closure should be replaced with _.

**Comment by at-thinhnguyen-2** [comment]
Total linter violations in pull request: 3
Serious: 0
Warnings: 3

**Comment by at-thinhnguyen-2** [comment]
:+1: There are no linter violations.

File: FinalProject/View/Detail/DetailCell/DetailTableViewCell.swift [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+import UIKit
+
+final class DetailTableViewCell: UITableViewCell {
+
+    // MARK: - IBOulets
+    @IBOutlet private weak var newCasesLabel: UILabel!
+    @IBOutlet private weak var newDeathsLabel: UILabel!
+    @IBOutlet private weak var newRecoveredLabel: UILabel!
+    @IBOutlet private weak var newActiveLabel: UILabel!
+    // MARK: - Properties
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Detail/DetailCell/DetailTableViewCell.swift:18]: line down

File: FinalProject/View/Detail/DetailCell/DetailTableViewCell.swift [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+    // MARK: - Private methods
+    private func updateView() {
+        guard let cases = viewModel?.newConfirmed else { return }
+        newCasesLabel.text = String(cases)
+        guard let deaths = viewModel?.newDeadth else { return }
+        newDeathsLabel.text = String(deaths)
+        guard let recovered = viewModel?.newRecovered else { return }
+        newRecoveredLabel.text = String(recovered)
+        guard let active = viewModel?.newActive else { return }
+        newActiveLabel.text = String(active)
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Detail/DetailCell/DetailTableViewCell.swift:34]: dùng if let

File: FinalProject/View/Detail/DetailViewController.swift [hunk 30]
```diff
@@ -0,0 +1,53 @@
...
+    // MARK: - Life cycle
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        title = viewModel.title
+    }
+
+    // MARK: - Override methods
+    override func setUpUI() {
+        let nib = UINib(nibName: "DetailTableViewCell", bundle: .main)
+        tableView.register(nib, forCellReuseIdentifier: cellIdentifier)
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Detail/DetailViewController.swift:29]: tableView.register(nibWithCellClass: DetailTableViewCell.self)

File: FinalProject/View/Detail/DetailViewController.swift [hunk 31]
```diff
@@ -0,0 +1,53 @@
...
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier, for: indexPath) as? DetailTableViewCell else { return UITableViewCell() }
+        return cell
+    }
+}
+
+// MARK: - Extension UITableViewDelegate
+extension DetailViewController: UITableViewDelegate {
+
+}
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Detail/DetailViewController.swift:53]: chưa dùng thì xoá đi nha

File: FinalProject/View/Detail/DetailViewController.swift [hunk 32]
```diff
@@ -0,0 +1,53 @@
...
+
+// MARK: - Extension UITableViewDataSource
+extension DetailViewController: UITableViewDataSource {
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return 5
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier, for: indexPath) as? DetailTableViewCell else { return UITableViewCell() }
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Detail/DetailViewController.swift:45]: let cell = tableView.dequeueReusableCell(withClass: DetailTableViewCell.self)

File: FinalProject/View/Detail/DetailViewController.swift [hunk 33]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    // MARK: - Private methods
+}
+
+// MARK: - Extension UITableViewDataSource
+extension DetailViewController: UITableViewDataSource {
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return 5
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Detail/DetailViewController.swift:41]: viết hàm ở viewModel

File: FinalProject/View/Detail/DetailViewModel.swift [hunk 34]
```diff
@@ -0,0 +1,43 @@
...
+                completion( .failure(error))
+            case .success(let result):
+                self.lastItem = result
+                completion( .success)
+            }
+        }
+    }
+
+    func viewModelForCirceChart(at indexPath: IndexPath) -> DetailCellModel {
+        let item = lastItem
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Detail/DetailViewModel.swift:39]: ko cần tạo thêm biến item

File: FinalProject/View/Login/LoginViewController.swift [hunk 35]
```diff
@@ -14,73 +14,73 @@ import FBSDKLoginKit
...
     // MARK: - Properties
     private var appleButton: ASAuthorizationAppleIDButton!
-    
+
     // MARK: - Life cycle
     override func viewDidLoad() {
         super.viewDidLoad()
         configAppleButton()
         configGoogleButton()
         GIDSignIn.sharedInstance()?.presentingViewController = self
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Login/LoginViewController.swift:29]: bỏ trong hàm `configGoogleButton()`

File: FinalProject/View/Search/Cell/SearchTableViewCell.swift [hunk 36]
```diff
@@ -0,0 +1,42 @@
...
+        countryCodeLabel.text = viewModel?.countryCode
+        guard let numberCase = viewModel?.totalConfirmed else { return }
+        numbercaseLabel.text = String(numberCase)
+        guard let numberDeadth = viewModel?.totalDeadth else { return }
+        numberDeadthLabel.text = String(numberDeadth)
+        guard let numberRecovered = viewModel?.totalRecovered else { return }
+        numberRecoveredLabel.text = String(numberRecovered)
+    }
+
+    @IBAction private func followButtonTouchUpInside(_ sender: UIButton) {
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/Cell/SearchTableViewCell.swift:40]: chưa có action thì để
// TODO: - ....

File: FinalProject/View/Search/SearchViewModel.swift [hunk 37]
```diff
@@ -0,0 +1,48 @@
...
+//  Copyright © 2020 Thinh Nguyen X. All rights reserved.
+//
+
+import Foundation
+
+final class SearchViewModel {
+
+    // MARK: - Properties
+    var countries: [Country] = []
+    var filter: [Country] = []
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/SearchViewModel.swift:15]: sao phải tạo 2 biến khác nhau

**Review by at-thinhnguyen-2 — COMMENTED** [review]
@at-tamnguyen7 please check comment

File: FinalProject/View/Search/SearchViewModel.swift (hunk 37 above)
Comment (at-tamnguyen7) [reply FinalProject/View/Search/SearchViewModel.swift:15]: Một mảng chứa dữ liệu , một mảng chưa ds search a ơi

**Comment by at-thinhnguyen-2** [comment]
Total linter violations in pull request: 3
Serious: 0
Warnings: 3

**Comment by at-thinhnguyen-2** [comment]
:+1: There are no linter violations.

File: FinalProject/View/Search/SearchViewModel.swift (hunk 37 above)
Comment (at-thinhnguyen-2) [reply FinalProject/View/Search/SearchViewModel.swift:15]: @at-tamnguyen7 filter  thêm `s`

File: FinalProject/Model/Schema/DayOneCountry.swift [hunk 38]
```diff
@@ -0,0 +1,47 @@
...
+//    var deathRate: Double {
+//        return (Double(deaths) / Double(confirmed)) * 100
+//    }
+//    var recoverRate: Double {
+//        return (Double(recovered) / Double(confirmed)) * 100
+//    }
+//    var activeRate: Double {
+//        return Double()
+//    }
+}
```
Comment (at-thinhnguyen-2) [inline FinalProject/Model/Schema/DayOneCountry.swift:47]: remove it

File: FinalProject/View/Login/LoginViewController.swift [hunk 39]
```diff
@@ -89,26 +89,27 @@ final class LoginViewController: ViewController {
...
-        return self.view.window!
+        guard let window = self.view.window else { return ASPresentationAnchor() }
+        return window
     }
-    
+
     func authorizationController(controller: ASAuthorizationController, didCompleteWithAuthorization authorization: ASAuthorization) {
         if let appleIDCredential = authorization.credential as? ASAuthorizationAppleIDCredential {
             guard let appleIDToken = appleIDCredential.identityToken else {
                 print("Unable to fetch identity token")
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Login/LoginViewController.swift:101]: Xoá print

File: FinalProject/View/Login/LoginViewController.swift [hunk 40]
```diff
@@ -89,26 +89,27 @@ final class LoginViewController: ViewController {
...
                 return
             }
-            
-            guard let idTokenString = String(data: appleIDToken, encoding: .utf8) else {
+
+            guard String(data: appleIDToken, encoding: .utf8) != nil else {
                 print("Unable to serialize token string from data: \(appleIDToken.debugDescription)")
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Login/LoginViewController.swift:106]: Xoá print

File: FinalProject/View/Search/Cell/SearchTableViewCell.swift [hunk 41]
```diff
@@ -0,0 +1,43 @@
...
+        didSet {
+            updateView()
+        }
+    }
+
+    // MARK: - Private methods
+    private func updateView() {
+        countryNameLabel.text = viewModel?.countryName
+        countryCodeLabel.text = viewModel?.countryCode
+        guard let numberCase = viewModel?.totalConfirmed else { return }
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/Cell/SearchTableViewCell.swift:32]: dùng if let thây cho guard let

File: FinalProject/View/Search/SearchViewController.swift [hunk 42]
```diff
@@ -8,23 +8,102 @@
...
     override func viewDidLoad() {
         super.viewDidLoad()
+        loadData()
+        configSearchController()
+    }
+
+    // MARK: - Override methods
+    override func setUpUI() {
+        let nib = UINib(nibName: "SearchTableViewCell", bundle: .main)
+        tableView.register(nib, forCellReuseIdentifier: cellIdentifier)
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/SearchViewController.swift:32]: dùng : 
tableView.register(nibWithCellClass: SearchTableViewCell.self)

File: FinalProject/View/Search/SearchViewController.swift [hunk 43]
```diff
@@ -8,23 +8,102 @@
...
+        searchController.searchResultsUpdater = self
+    }
+
+    private func loadData() {
+        viewModel.getData() { result in
+            switch result {
+            case .success:
+                DispatchQueue.main.async {
+                    self.tableView.reloadData()
+                }
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/SearchViewController.swift:51]: không cần dùng: DispatchQueue.main.async

File: FinalProject/View/Search/SearchViewController.swift [hunk 44]
```diff
@@ -8,23 +8,102 @@
...
+            case .failure(let error):
+                DispatchQueue.main.async {
+                    self.alert(error: error)
+                }
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/SearchViewController.swift:55]: DispatchQueue.main.async

File: FinalProject/View/Search/SearchViewController.swift [hunk 45]
```diff
@@ -8,23 +8,102 @@
...
+            }
+        }
+    }
+
+   private func filterCountry(for searchText: String) {
+            if searchText.isEmpty {
+                viewModel.filter = viewModel.countries
+            } else {
+                viewModel.filter = viewModel.countries.filter { country in
+                    return
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/SearchViewController.swift:65]: sao return chổ này

File: FinalProject/View/Search/SearchViewController.swift [hunk 46]
```diff
@@ -8,23 +8,102 @@
...
+// MARK: - Extension UITableViewDataSource
+extension SearchViewController: UITableViewDataSource {
 
-        // Do any additional setup after loading the view.
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return viewModel.numberOfRowInSection()
     }
 
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier, for: indexPath) as? SearchTableViewCell else { return UITableViewCell() }
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/SearchViewController.swift:81]: let cell = tableView.dequeueReusableCell(withClass: SearchTableViewCell.self, for: indexPath)

File: FinalProject/View/Search/SearchViewModel.swift [hunk 47]
```diff
@@ -0,0 +1,46 @@
...
+
+    // MARK: - Function
+    func getData(completion: @escaping APICompletion) {
+        Api.Path.Search.getAllDataInSearch() { result in
+            switch result {
+            case .failure(let error):
+                completion( .failure(error))
+            case .success(let result):
+                self.countries.append(contentsOf: result)
+                self.filter = self.countries
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/SearchViewModel.swift:25]: bỏ self

File: FinalProject/View/Search/SearchViewController.swift [hunk 48]
```diff
@@ -8,23 +8,102 @@
...
+    // MARK: - Private methods
+    private func configSearchController() {
+        title = titleSearch
+        navigationItem.searchController = searchController
+        searchController.obscuresBackgroundDuringPresentation = false
+        searchController.searchResultsUpdater = self
+    }
+
+    private func loadData() {
+        viewModel.getData() { result in
```
Comment (at-thinhnguyen-2) [inline FinalProject/View/Search/SearchViewController.swift:46]: viewModel.getData() { [weak self] result in
guard let this = self else { return }
....

**Review by at-thinhnguyen-2 — COMMENTED** [review]
@at-tamnguyen7 please check comment

**Comment by at-thinhnguyen-2** [comment]
Total linter violations in pull request: 12
Serious: 0
Warnings: 12

**Comment by at-thinhnguyen-2** [comment]
:+1: There are no linter violations.

**Review by at-thinhnguyen-2 — APPROVED** [review]
LGTM!

**Pull request merged**